feat(onboarding): Add OnboardingViewModel tests and improve effect handling

Mostafa-alsaygh opened the pull request:
# What does this PR do?

this PR introduces comprehensive unit tests for `OnboardingViewModel`, covering various scenarios like page changes, scrolling, and navigation.

Key changes:
- Added `OnboardingViewModelTest.kt` with tests for all public methods and state updates.
- Refactored `onboardingScreen.kt` to use `Listen` for handling `OnboardingEffect`, simplifying the effect handling logic and removing `LaunchedEffect`.
- Removed unnecessary logging in `OnboardingViewModel` for `onboardingFinished`.
- Updated `welcomeScreen.kt` to use `onNavigateAsGuest` instead of `onNavigateContinue` for better clarity.
- Included `kotlin-test` dependency in `presentation/build.gradle.kts` for unit testing.
- Updated `onboardingNavGraph.kt` to reflect the name change from `onNavigateContinue` to `onNavigateAsGuest`.

Mostafa-alsaygh committed:
feat(onboarding): Add OnboardingViewModel tests and improve effect handling

This commit introduces comprehensive unit tests for `OnboardingViewModel`, covering various scenarios like page changes, scrolling, and navigation.

Key changes:
- Added `OnboardingViewModelTest.kt` with tests for all public methods and state updates.
- Refactored `onboardingScreen.kt` to use `Listen` for handling `OnboardingEffect`, simplifying the effect handling logic and removing `LaunchedEffect`.
- Removed unnecessary logging in `OnboardingViewModel` for `onboardingFinished`.
- Updated `welcomeScreen.kt` to use `onNavigateAsGuest` instead of `onNavigateContinue` for better clarity.
- Included `kotlin-test` dependency in `presentation/build.gradle.kts` for unit testing.
- Updated `onboardingNavGraph.kt` to reflect the name change from `onNavigateContinue` to `onNavigateAsGuest`.

yusufnasserdev commented on presentation/build.gradle.kts:
```diff
@@ -67,4 +67,5 @@ dependencies {
 
     ksp(libs.bundles.hilt.ksp)
     implementation(libs.bundles.hilt.runtime)
+    testImplementation(kotlin("test"))
```
this is redundant, plz revert it.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/welcome/onboarding/OnboardingViewModel.kt:
```diff
@@ -58,7 +57,6 @@ class OnboardingViewModel @Inject constructor(
     fun onboardingFinished() {
         viewModelScope.launch(Dispatchers.IO) {
             runCatching { appPreferencesService.setOnBoardingShown() }
-                .onFailure { Log.e("OnboardingViewModel", "onboardingFinished: ", it) }
```
Why remove the log?

Mostafa-alsaygh replied:
why is it there ?
in final code it should be clean dont have  a old logs

yusufnasserdev reviewed (changes requested):
Kindly check failed test below:

OnboardingViewModelTest > when onboardingFinished is called, preferences service should be invoked FAILED
    java.lang.AssertionError at OnboardingViewModelTest.kt:255

Mostafa-alsaygh committed:
Refactor: Rename OnboardingEffect.NavigateToWelcome to OnWelcomeNavigation

This commit renames the `OnboardingEffect.NavigateToWelcome` to `OnboardingEffect.OnWelcomeNavigation` for better clarity.

Additionally, it removes an unused `testImplementation(kotlin("test"))` dependency from the `presentation/build.gradle.kts` file.

Mostafa-alsaygh committed:
Merge branch 'develop' into add/onboarding-viewmodel-testing

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/welcome/onboarding/welcomeScreen.kt

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Refactor: Use tryToExecute for preference setting

Refactors the `onboardingFinished` and `setOnBoardingShown` functions in `OnboardingViewModel` to use the `tryToExecute` extension function for setting the onboarding shown preference. This improves error handling and logging.

Mostafa-alsaygh committed:
Merge branch 'develop' into add/onboarding-viewmodel-testing

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/welcome/onboarding/OnboardingViewModel.kt

Mostafa-alsaygh committed:
Test: Improve OnboardingViewModelTest stability

This commit addresses potential flakiness in `OnboardingViewModelTest` by:

- Introducing `advanceTimeBy(1000)` and `advanceUntilIdle()` in tests involving `onboardingFinished()` and `navigateToWelcome()`. This ensures that IO operations (specifically, `appPreferencesService.setOnBoardingShown()`) have sufficient time to complete before assertions are made.
- Adding an alternative test case (`onboardingFinished is called, preferences service should be invoked - with timeout`) that utilizes `coVerify(timeout = 2000)` as another way to handle asynchronous operations, providing more robust verification.

These changes enhance the reliability of the tests by better managing asynchronous operations within the `runTest` environment.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Merge branch 'develop' into add/onboarding-viewmodel-testing

# Conflicts:
#	app/src/main/java/com/london/app/navigation/graph/onboardingNavGraph.kt
#	presentation/src/main/java/com/london/presentation/feature/welcome/onboarding/OnboardingScreen.kt

Mostafa-alsaygh committed:
test(OnboardingViewModel): Advance time in scrollNext test

The `scrollNext` function in `OnboardingViewModel` involves a coroutine scope for scrolling the Pager. This commit adds `advanceTimeBy(1000)` to the corresponding test case to ensure the coroutine has enough time to complete, preventing potential flakiness.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
test(OnboardingViewModel): Remove redundant comments

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/add/onboarding-viewmodel-testing' into add/onboarding-viewmodel-testing

Mostafa-alsaygh committed:
test(OnboardingViewModel): Replace `advanceTimeBy` with `withTimeout` and `coVerify` with `timeout`

This commit refactors tests in `OnboardingViewModelTest` to improve reliability and readability.

- `advanceTimeBy(1000)` has been replaced with `withTimeout(2000)` when asserting navigation effects. This provides a more explicit timeout mechanism for operations that might take longer.
- For `coVerify` calls that interact with `appPreferencesService`, a `timeout = 2000` parameter has been added directly to the `coVerify` function. This ensures the verification waits for the asynchronous preference setting operation to complete.

These changes make the tests more robust by explicitly handling potential timeouts for asynchronous operations, particularly those involving I/O like preference updates and navigation.

Mostafa-alsaygh committed:
Merge branch 'develop' into add/onboarding-viewmodel-testing

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/category/tvshow/TvShowCategoryScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/actor/info/gallery/ActorsGalleryScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/home/trending/actor/TrendingActorsScreen.kt

Mostafa-alsaygh committed:
test(OnboardingViewModel): Remove unnecessary timeout in scrollNext test

Mostafa-alsaygh committed:
test(OnboardingViewModel): Use advanceUntilIdle in navigateToWelcome test

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/welcome/onboarding/OnboardingEffect.kt:
```diff
@@ -1,6 +1,6 @@
 
 sealed interface OnboardingEffect {
-    data object NavigateToWelcome : OnboardingEffect
+    data object OnWelcomeNavigation : OnboardingEffect
```
Effect names don't start with `on`, they define an effect, hence they are a noun.

It should be `WelcomeNavigation`

Mostafa-alsaygh replied:
i would name it like this but i check another effects in the project and see this approach so i followed it

yusufnasserdev commented on presentation/src/test/java/com/london/presentation/feature/onboarding/OnboardingViewModelTest.kt:
Can you kindly clarify further why are you using `coVerify` with `timeout` argument?

Mostafa-alsaygh replied:
On a slower CI runner fewer cores or threads, the coroutine may not run before the test checks the mock.
This race condition is why you see failures only in CI/CD.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/add/onboarding-viewmodel-testing' into add/onboarding-viewmodel-testing

Mostafa-alsaygh committed:
Refactor: Rename OnboardingEffect.OnWelcomeNavigation to WelcomeNavigation
